proposals/0023-asynchronous-program-execution-and-broadcast.md

```diff
@@ -89,26 +90,63 @@ If a validator doesn't have the builder's UserBlock, the validator
 doesn't vote on the proposed block and tries to repair the UserBlock.
 That fork is ignored for fork choice.
 
-Otherwise the validator evaluates the forks as usual. 
+Otherwise the validator evaluates the forks as usual.
 
-### UserBlock schedule
+### UserBlock builder schedule
 
-There are 2 UserBlockSlots per slot. Builders produce 1 UserBlock
-per UserBlockSlot. So if the cluster is configured with 3 Builders,
-each leader slot would see a total of 6 UserBlocks per slot.  3 in
-the first half of the leaders slot, and 3 in the second half.
+Each block is sloted to support N user blocks. Some of these builders
+are randomly scheduled, some of these are persistent. For example,
+there are 10 UserBlock builders per network block, that means 10
+builders are each assigned 10% of the shreds and 10% of the compute
+available to the block.
 
-The current leader schedule rotates each leader every 4 slots, but
-builders could be rotated at a different schedule then the leaders
-and rotation could be staggered.
+#### Randomized UserBlock builders
 
-### UserBlock Compute Limits
+Randomized UserBlock builders are assigned at the same time as the
+leader schedule is created based on stake weighted leader distribution.
+
+The benefit to more than 1 random UserBlock builder per leader
+block is that they are likely geographically distributed and users
+will be able to pick the closest one.
+
+The downside is that resource management becomes more complex. Each
+builder has 1/N compute and shred capacity and users have no idea
+which one is saturated when sending their transaction.  It's likely
+that priority fee floor will be different at each UserBlock builder.
+
+#### Persistent UserBlock slots
+
+Persistent UserBlock slots for the epoch are auctioned of to the top
+N bidders who burn the most lamports.
+
+In the first half of the epoch each builder deposits the lamports
+they are planning on burning, in the second half the builders may
+withdraw excess lamports.  The top N builders are assigned the slots
+in a dutch auction according to their remaining bids. If there are
```

Comment:
Can you clarify the auction process here? I would have assumed an auction process where we take the top `N` bids with time-priority, not a dutch auction.

Reply:
There are N slots to auction, so a Dutch auction makes the most sense.  Once bid could get 70% of the block and the other could get 30%

Reply:
That doesn't sound like a Dutch auction as I understand them, so maybe I'm just getting confused by the terminology.

It'd help to give an example in the document to clarify.

Reply:
Welp, since I'm the one who was confused I looked it up: https://www.investopedia.com/terms/d/dutchauction.asp

Seems to be an overloaded term, I was thinking of it as the first way they describe; i.e. a descending price auction, first bid wins.

I'm guessing that you mean the style they describe in the "Understanding Dutch Auctions for Initial Public Offerings (IPOs)" section:
> Once all the bids are submitted, the allotted placement is assigned to the bidders from the highest bids down, until all the allotted shares are assigned. However, the price that each bidder pays is based on the lowest price of all the allotted bidders, or essentially the last successful bid.

Reply:
The auction is run for the next epoch. So it takes a full epoch to finish.

Reply:
What's the most fair auction if the attacker can always place the last bid?

Reply:
> What's the most fair auction if the attacker can always place the last bid?

What do you mean by attacker in this case? A node trying to become a persistent builder by gaming the auction?

If we're worried about nodes waiting to submit bids until they see everyone elses, we could run a sealed-bid auction.

Reply:
Yea. Sealed bids might be better.  There is little harm in cancelled bids since the unused capacity would be scheduled randomly

Reply:
Alternative to an auction would be extending staking to pick persistent builders.  That choice might be necessary if slashing for duplicate UserBlocks is necessary

Reply:
Slashing duplicate UserBlocks shouldn't ever be necessary, since the leader would have picked 1 of the proposed blocks and the network would either accept or reject the leaders fork.  Everyone with inconsistent data can dump it and repair the right version.

Reply:
i would advocate for a second-price sealed-bid auction, one where winner pays the bid of the second place bidder

Reply:
> Alternative to an auction would be extending staking to pick persistent builders. That choice might be necessary if slashing for duplicate UserBlocks is necessary

how do you prove duplicate user blocks?

Reply:
Same as you prove duplicate blocks.  Last shred is different